### ISSUE TITLE
`secrets` not in `sys.path` when using `--venv` 

### ISSUE THREAD

**Issue by yjabri** [pantsbuild/pex#2235]
# Overview
This arose from a discussion on [slack](https://pantsbuild.slack.com/archives/C046T6T9U/p1694200846761159). The repo https://github.com/yjabri/debug-pex-import describes how to reproduce the bug in the `README.md`.

The culprit appears to be [pex_bootstrapper.py#L672](https://github.com/pantsbuild/pex/blob/main/pex/pex_bootstrapper.py#L672). Just prior to the line, in a debugger I can successfully `import secrets`. (A cause of `flask` failing to import). However just after the line, in a new debugger session `import secrets` raises a `ModuleNotFoundError` exception.

## Some details

The line [https://github.com/pantsbuild/pex/blob/main/pex/pex_bootstrapper.py#L672](pex_bootstrapper.py#L672 ) converts my `sys.path` from

```python
['/home/yjabri/Projects/debug-pex-import',
'/home/yjabri/.pyenv/versions/3.10.13/lib/python310.zip',
'/home/yjabri/.pyenv/versions/3.10.13/lib/python3.10',
'/home/yjabri/.pyenv/versions/3.10.13/lib/python3.10/lib-dynload',
'',
'/home/yjabri/.local/lib/python3.10/site-packages',
'/home/yjabri/.pyenv/versions/3.10.13/lib/python3.10/site-packages',
'dist/.pex' ]
```

to

```python
['/home/yjabri/.pex/venvs/6c8eaec5b07b6fe5d8c257cd72d1297dfafdfc8e/779eb2cc0ca9e2fdd204774cbc41848e4e7c5055/lib/python3.10/site-packages',
'dist/.pex']
 ```
 
 My understanding is that this doesn't include standard library which would be in  `'/home/yjabri/.pyenv/versions/3.10.13/lib/python3.10'`
 
For reference, `python dist/.pex -c "import sys; print(sys.path)"` outputs

```python
['',
'/home/yjabri/.pex/venvs/6c8eaec5b07b6fe5d8c257cd72d1297dfafdfc8e/779eb2cc0ca9e2fdd204774cbc41848e4e7c5055',
'/home/yjabri/.pyenv/versions/3.10.13/lib/python310.zip',
'/home/yjabri/.pyenv/versions/3.10.13/lib/python3.10',
'/home/yjabri/.pyenv/versions/3.10.13/lib/python3.10/lib-dynload',
'/home/yjabri/.pex/venvs/6c8eaec5b07b6fe5d8c257cd72d1297dfafdfc8e/779eb2cc0ca9e2fdd204774cbc41848e4e7c5055/lib/python3.10/site-packages']
```

## Why does this happen?

 If I set

```python
sys.path = [ 
    '/home/yjabri/.pyenv/versions/3.10.13/lib/python3.10',
    '/home/yjabri/.pex/venvs/6c8eaec5b07b6fe5d8c257cd72d1297dfafdfc8e/779eb2cc0ca9e2fdd204774cbc41848e4e7c5055/lib/python3.10/site-packages',
    'dist/.pex'
]
```

I can `import secrets` without a `ModuleNotFoundError`.

In general, if I create a `venv` manually via

```
python -m venv venv
```

and run `venv/bin/python -c "import sys; print(sys.path)"` I see 

```python
['', '/home/yjabri/.pyenv/versions/3.10.13/lib/python310.zip', '/home/yjabri/.pyenv/versions/3.10.13/lib/python3.10', '/home/yjabri/.pyenv/versions/3.10.13/lib/python3.10/lib-dynload', '/tmp/foobar/venv/lib/python3.10/site-packages']
```

I get `ModuleNotFoundError` by running 

```
venv/bin/python -c "import sys; sys.path = sys.path[3:]; import secrets"
```

but not

```
venv/bin/python -c "import sys; sys.path = sys.path[2:]; import secrets"
```

does not raise the exception. If I replace `secrets` with other standard library packages like `http` or `random` I get the same `ModuleNotFoundError` exception.

**Comment by yjabri** [comment]
If I make the modification

```python
with ENV.patch(PEX_ROOT=pex_info.pex_root):
    if not execute:
        for location in _activate_pex(entry_point, pex_info, venv_dir=venv_dir):
            from pex.third_party import VendorImporter

            VendorImporter.install(
                uninstallable=False, prefix="__pex__", path_items=["."], root=location
            )

        # Include compatible interpreter for python standard library
        interpreter_test = InterpreterTest(entry_point=entry_point, pex_info=pex_info)
        try:
            target = find_compatible_interpreter(interpreter_test=interpreter_test)
        except UnsatisfiableInterpreterConstraintsError as e:
            die(str(e))
        sys.path.insert(0, "{}/lib/python{}".format(target.prefix, target.python))

        return
```

I can `import flask` without a `ModuleNotFoundError` exception after

Would a PR be welcomed? I'd probably need a little time to learn how to run tests in this repo as I've never used `tox` before.

**Comment by yjabri** [comment]
Awesome. In the PR I'd create a test in `pext/tests/integration`  for this issue. Removing all the unnecessary  details about `ipython` and `jupyter` this can be distilled to the following


### Expected Behavior
The following is expected behavior. Run

```bash
rm -fr output.txt
dist/pex Flask -o flask.pex
python script.py
cat output.txt
```

where `script.py` is

```python
import sys
error = ""

try:
    sys.path.append("flask.pex")
    import __pex__
    import flask
except Exception as e:
    error = str(e)

with open("output.txt", 'w') as f:
    f.write(error)
```

and observe the file is empty. Note I excluded `--venv` when building the `pex`.  It's worth noting that `python --version` prints
`Python 3.9.18`

### Unexpected Behavior
To reproduce the bug run 

```bash
rm -fr output.txt
dist/pex Flask -o flask.pex
python script.py
cat output.txt
```

and observe `output.txt` contains the line `No module named 'contextvars`.

### Writing a Test
To write a test, I'm going to need to use `subprocess` to orchestrate the following 

1. Create a pex from source. `test_issue_1872.py` looks like a good reference
2. Write the above script to a tmpdir
3. Execute it with python (I think I ought to use `sys.executable`) for this?
4. Check `output.txt` and assert it's empty.

Does that sound good?

**Comment by jsirois** [comment]
That sounds about right. You should pin the flask dep so that CI is more stable over time and the version of Flask you pin will necessitate adding a `@pytest.mark.skipif` that skips the test for unsupported Pythons, likely `sys.version_info[:2] < (3, 7)` (or 3.8).

**Comment by yjabri** [comment]
Is the intention to strip all of the following?

```
/home/yjabri/.pyenv/versions/3.9.18/lib/python39.zip
/home/yjabri/.pyenv/versions/3.9.18/lib/python3.9
/home/yjabri/.pyenv/versions/3.9.18/lib/python3.9/lib-dynload
```

If I try something like

```python
import sys
sys.path[:] = [
    "/home/yjabri/.pyenv/versions/3.9.18/lib/python39.zip",
    "/home/yjabri/.pyenv/versions/3.9.18/lib/python3.9",
    "/home/yjabri/.pyenv/versions/3.9.18/lib/python3.9/lib-dynload",
]
import secrets
```

It only works if I have (`/home/yjabri/.pyenv/versions/3.9.18/lib/python39.zip` or `/home/yjabri/.pyenv/versions/3.9.18/lib/python3.9`) and `/home/yjabri/.pyenv/versions/3.9.18/lib/python3.9/lib-dynload`. 

The specific paths seem like an implementation detail specific to whatever installed the interpreter being used.

For example I need different paths if using `/usr/bin/python3.9`. I haven't even checked what these would be on Mac (or Windows).

Would it make sense to capture `sys.path` near the beginning of `BOOTSTRAP_ENVIRONMENT` and ensure we don't strip necessary paths like the ones listed above? Would this break other uses of  `__pex__`?

I'm going to thinking about this, but thought I'd check  base and make sure I'm on the right path towards solving this.

**Comment by jsirois** [comment]
As I said above (https://github.com/pantsbuild/pex/issues/2235#issuecomment-1716808615) and you originally sketched above (https://github.com/pantsbuild/pex/issues/2235#issuecomment-1712726080) I think `sys.path.insert(0, ...)` is the right thing to do and probably the only needed fix. The thing to insert is the same thing ~inserted today, namely `venv.site_packages_dir`.

**Comment by jsirois** [comment]
Maybe this makes the situation more clear - I have absolutely no clue why I did this crazy slice nor how it passed review:
https://github.com/pantsbuild/pex/blob/5c0c29c192bd99335b5a1ab5b68a0f8556cd97aa/pex/pex_bootstrapper.py#L672

The only operations that make sense are to either prepend the venv PEX site-packages to the existing `sys.path` or else append to the existing `sys.path`. Nuking parts of the existing `sys.path` is insane. Somehow though I proceeded to do that and get it checked in.
